v0.02 complete

JodyLS27 opened the pull request:
I have completed phase two which includes correct collision with the window, AABB Collision with the target, a score counter for the player, and improved movement of the player object.

JodyLS27 committed:
- Adding a completed CMakeLists.txt with a FetchContent from SFML \n- Added an ignor for the ext folder.

JodyLS27 committed:
Setup main.cpp with window render and added the player.cpp and player.hpp files.

JodyLS27 committed:
Adding the player source and header file logic.

JodyLS27 committed:
Added some more Functionality to the Player Header and source file.

JodyLS27 committed:
Added additional logic to the player class

JodyLS27 committed:
Adding logic to the set_direction() function.

JodyLS27 committed:
- Added Player Movement structure
- Added main source cleanup and clarity.

JodyLS27 committed:
- Applied fixes from Dev Branch pull request comments.

JodyLS27 committed:
Added improved player movment and class design.

JodyLS27 committed:
Added my first basic implementation of the AABB collision method.
In this test I have done a simple check on the right hand side of the target and
switching colours if we move into the target or to the left hand side.

JodyLS27 committed:
- Finalized the AABB collision detection
- incremented the player score on collision with target
- Moved the target on collision

swagween commented on src/player.cpp:
```diff
@@ -0,0 +1,117 @@
+
+
+player::Player::Player(int score, int speed)
+	: m_score(score), m_speed(speed), m_width(25), m_height(25)
```
you could set `m_width` and `m_height` using optional constructor arguments instead of hard-coding it

karnkaul replied:
Even better: set them in the class definition itself

swagween commented on src/player.cpp:
```diff
@@ -0,0 +1,117 @@
+	m_shape.setFillColor(color);
+}
+
+void player::Player::set_score()
```
I would rename this function to `increment_score()`

swagween commented on src/player.cpp:
```diff
@@ -0,0 +1,117 @@
+{
+	sf::Vector2f ret_position = m_position + (m_direction * delta_time) * m_speed;
+
+	if (ret_position.x < (m_width * 0.5f)) { ret_position.x = (m_width * 0.5f); }
```
consider writing a `get_center()` method so you don't have to keep writing `m_width * 0.5f`

swagween commented on src/player.hpp:
```diff
@@ -0,0 +1,44 @@
+
+namespace player
+{
+	class Player
```
`namespace` and its highest level code block (in this case, `class Player {}`) should have the same indentation

JodyLS27 replied:
Hey there.
Yea these are the indentations from Visual Studio... I need to look into a different styling as this is what it gives me even if I put it with the same indentations.

JodyLS27 replied:
@karnkaul Should I be using Clang formatting ?

swagween commented on src/target.cpp:
```diff
@@ -0,0 +1,87 @@
+	sf::Vector2f position{ static_cast<float>(random_val(0, window_size.x)), static_cast<float>(random_val(0, window_size.y)) };
+
+	// Bounds checking.
+	if (position.x < m_radius) { position.x = m_radius; }
```
this is sorta duplicated code; in the future, you could consider writing some kind of `collider` component that has a method with this code, and then give both `target` and `player` a member `Collider` and check window collision against that.

JodyLS27 replied:
Hey :)

So I am going to be implementing a collision class that will deal with this so I can just pass objects in and check their collision state. It will help with redundancy and clean it up a bit 😄

swagween commented on src/main.cpp:
```diff
@@ -1,6 +1,108 @@
-    return 0;
-}
\ No newline at end of file
+	player::Player player(0, 100.0f);
```
I thought this should to be initialized with `{}` but I'm not really sure if it makes a difference

JodyLS27 replied:
Yea I did some looking and your right, Will adjust thank you 👍

swagween commented on src/main.cpp:
```diff
@@ -1,6 +1,108 @@
+			}
+		}
+
+		logic(player, target, event, window_size, delta_time);
```
I think you want to handle events separately from logic. `Player::event_handler()` should be called whilst the window is polling the event queue. Logic can be applied after all events are processed. I'm not sure about the consequences of doing it your way, and it's possible that it's totally fine, but it seems like it could lead to some unexpected behavior.

karnkaul replied:
+1
Store the state changes in the event loop, then respond to those changes later in the tick / update loop.

JodyLS27 replied:
Hey 😃 
So If I place the player.event_handler(event) in the while loop for the pollEvent(), I get unexpected behavior where the player object starters each time a different key is pressed plus there is a strong lag due to the loop.
So placing it outside helped smooth out the movement.

Not sure if I should go back to default and just have a case key switch event in the main window ?

swagween replied:
the reason for that is probably this part of your `set_direction()` function.
```cpp
m_direction = { 0.0f, 0.0f };
if (m_move_up) { m_direction = { 0.0f, -1.0f }; }
if (m_move_left) { m_direction = { -1.0f, 0.0f }; }
if (m_move_down) { m_direction = { 0.0f, 1.0f }; }
if (m_move_right) { m_direction = { 1.0f , 0.0f }; }
```
Think about what will happen here if the player presses and holds the `right` key, then also presses the `up` key. The way you have it here, there is a different priority for directions simply because of how they're ordered in the code. `right` beats everything because it's checked last, and `m_direction` is overwritten regardless of the other movement states.

Either way, you should switch to event handling inside the loop and troubleshoot the bugs from there. It's weird that you get a lag like you said, but I highly doubt it's from event polling because that happens pretty much instantly.

JodyLS27 replied:
Cool, yea I will look into this further, Thank you.

karnkaul commented on src/main.cpp:
```diff
@@ -1,6 +1,108 @@
+	using mil = std::chrono::milliseconds;
+	using flt = std::chrono::duration<float>;
+
+	auto end_clock = std::chrono::time_point_cast<mil>(clock::now());
```
Why is this needed?

JodyLS27 replied:
Hey Karn, 
Are you referring to the float variable?

karnkaul replied:
I see that you're assigning to it inside the main loop, that's cool then

karnkaul commented on src/main.cpp:
```diff
@@ -1,6 +1,108 @@
+	{
+		sf::Event event;
+
+		auto current_clock = std::chrono::time_point_cast<mil>(clock::now());
```
The cast is unnecessary, you can just subtract two time points implicitly into a float duration.

JodyLS27 replied:
Hey Karn, 
I did this casting as to not have the extra floating points in the time, but if its not needed to cast, I will look into remove it.

karnkaul replied:
Why do you not want extra floating points in time? That will make the simulation inaccurate.

karnkaul replied:
This is still truncating decimals

karnkaul commented on src/player.cpp:
```diff
@@ -0,0 +1,117 @@
+	if (m_move_up) { m_direction = { 0.0f, -1.0f }; }
+	if (m_move_left) { m_direction = { -1.0f, 0.0f }; }
+	if (m_move_down) { m_direction = { 0.0f, 1.0f }; }
+	if (m_move_right) { m_direction = { 1.0f , 0.0f }; }
```
What about diagonals?

karnkaul commented on src/player.cpp:
```diff
@@ -0,0 +1,117 @@
+
+void player::Player::set_score()
+{
+	m_score++;
```
Use pre increment by default

karnkaul commented on src/player.cpp:
```diff
@@ -0,0 +1,117 @@
+	return m_height;
+}
+
+void player::Player::event_handler(sf::Event event)
```
Don't copy events

karnkaul commented on src/player.cpp:
```diff
@@ -0,0 +1,117 @@
+		if (event.key.code == sf::Keyboard::W || event.key.code == sf::Keyboard::Up) { m_move_up = false; }
+		if (event.key.code == sf::Keyboard::A || event.key.code == sf::Keyboard::Left) { m_move_left = false; }
+		if (event.key.code == sf::Keyboard::S || event.key.code == sf::Keyboard::Down) { m_move_down = false; }
+		if (event.key.code == sf::Keyboard::D || event.key.code == sf::Keyboard::Right) { m_move_right = false; }
```
Improve this mess of `if`s 😄

JodyLS27 replied:
Yup... On it :)

karnkaul commented on src/player.cpp:
```diff
@@ -0,0 +1,117 @@
+
+void player::Player::update_position(sf::Vector2f position)
+{
+	m_position = position;
```
Why is `m_position` a separate variable when `m_shape.getPosition()` is sufficient?

JodyLS27 replied:
Yea... I could do that 🤔 Thanks Karn

karnkaul commented on src/player.cpp:
```diff
@@ -0,0 +1,117 @@
+	m_shape.setPosition(position);
+}
+
+void player::Player::update_position(float delta_time)
```
Take `chrono::duration<float>` as `delta_time`; `float` doesn't indicate what the unit is.

karnkaul commented on src/player.cpp:
```diff
@@ -0,0 +1,117 @@
+	if (ret_position.x < (m_width * 0.5f)) { ret_position.x = (m_width * 0.5f); }
+	if (ret_position.x > (window_size.x - (m_width * 0.5f))) { ret_position.x = (window_size.x - (m_width * 0.5f)); }
+	if (ret_position.y < (m_height * 0.5f)) { ret_position.y = (m_height * 0.5f); }
+	if (ret_position.y > (window_size.y - (m_height * 0.5f))) { ret_position.y = (window_size.y - (m_height * 0.5f)); }
```
Again lots of `if`s

karnkaul commented on src/player.hpp:
```diff
@@ -0,0 +1,44 @@
+{
+	class Player
+	{
+	private:
```
You'd generally want to put `public` stuff first, as its part of the class' API, and `private` at the bottom.

karnkaul commented on src/target.cpp:
```diff
@@ -0,0 +1,87 @@
+
+int ::target::Target::random_val(int lo, int hi)
+{
+	auto engine = std::default_random_engine{ std::random_device{}() };
```
This is incorrect, you're creating a new `engine` on every call. You want to use the same engine that's initialized with a seed _once_ throughout the entire program.

karnkaul commented on src/target.cpp:
```diff
@@ -0,0 +1,87 @@
+}
+
+
+sf::Vector2f target::Target::get_position()
```
Missing `const`

karnkaul commented on src/target.cpp:
```diff
@@ -0,0 +1,87 @@
+	if (position.x < m_radius) { position.x = m_radius; }
+	if (position.x > window_size.x - (m_radius * 1.5f)) { position.x = window_size.x - (m_radius * 1.5f); }
+	if (position.y < m_radius) { position.y = m_radius; }
+	if (position.y > window_size.y - (m_radius * 1.5f)) { position.y = window_size.y - (m_radius * 1.5f); }
```
Lots of `if`s

karnkaul commented on src/target.cpp:
```diff
@@ -0,0 +1,87 @@
+
+bool target::Target::collision_hit(const sf::Vector2f player_position, const int player_width, const int player_height)
+{
+	/// <summary>
```
Doc comments go above the function / class, not inside

karnkaul commented on src/main.cpp:
```diff
@@ -1,6 +1,108 @@
+	sf::Font font;
+	sf::Text text;
+
+	if (!font.loadFromFile("C:/Users/jodys/OneDrive/Documents/Fonts/Vogue.ttf"))
```
Pretty sure this file isn't going to exist anywhere except your dev machine 😄

JodyLS27 committed:
Altered the version to reflect the current version of the project.

JodyLS27 committed:
Changing plyaer inputs
- Adjusting the event handling out of the player class and into the main function.
- Updated the player width and height to be optional values with usable defaults.
- Changed set_score() to increment_score() with a pre increment instead of a post increment.

JodyLS27 committed:
Player Movment Update
- Improved the event handling system to work with states.
- Added player member functions, (set_move_up, left, right, down) for the state changes.

JodyLS27 committed:
Event Handling Bug
- Fixed bug with event handling, Cause was left out break for case events

JodyLS27 committed:
Updated int variables
- Updated from int to uint16_t in for variables
- Updated the Ints in player.hpp and plyaer.cpp
- Added get_center member function, Might not need.

JodyLS27 committed:
C++ Commit Log
Updating Player

- Removed m_position member varaible and replaced with m_shape getter and setters for positional data.

- Updated class initializations in main from assignment to list initialization.

JodyLS27 committed:
Added Font file
- Added the font file to the project structure.
- Changed the font path to be relative to the current working directory.

JodyLS27 committed:
Fixed coding format

JodyLS27 committed:
Layout Improvments and Removing m_position
- Removed m_position from player header and source file
- cleaned up player code using m_position
- Removed m_position from target header and source file
- Cleaned up target functions using m_position
- Organised layout on player and target files.

JodyLS27 committed:
Updating Player
- Optimizing the set_direction() function

JodyLS27 committed:
Movement Update
- Fixed the endless movement, Character will now stop if nothing is being pressed.

JodyLS27 committed:
Debugging player movement
- Working on solving the multi input direction change.
- Player can accept multi inputs like left and tapping right to get a glitchy effect.

JodyLS27 committed:
Clean Inputs
- Looked into Solutions for only one direction at a time.
- Did some Debug logging to check the state of inputs.
- Did not like the results as it gave a laggy input feel.

JodyLS27 committed:
Improving movement system
- Added a check to see if the player can_change_direction().
- can_change_direction() only allows for stationary players to change direction
and non oposit directions (i.e: cannot hold left and tap right, etc).
- Moved all DocStrings to header files and out of source files.
- Added a DocString to player::can_change_direction().
- Updated the DocString target::collision_hit().

JodyLS27 committed:
Implemented diagonal movement
- Added Diagonal checkes in set_direction()

JodyLS27 committed:
Code clean up
- Removed all debugging checks for testing.
- Removed any redundent TODO's.

JodyLS27 committed:
Removed std::out
- Removed std::out from the file error check message for the text format file.

JodyLS27 committed:
Updated GitIgnore

JodyLS27 committed:
Merge branch 'dev' of https://github.com/JodyLS27/MTT into dev

JodyLS27 committed:
Quick cleanup on includes

JodyLS27 committed:
Adding logger header
- Added includes for base setup.

JodyLS27 committed:
Added logger.cpp file
- Added simple print to terminal logger function
- Added windows debugging logger function

JodyLS27 committed:
Updating files
- Added debugging logger function to the logger.hpp file
- Updated the CMakeLists with the new logger files

JodyLStocks committed:
Updating logger
- Changed WIN32 to _WIN32

JodyLStocks committed:
Updated random engine position
- Changed the random engine to only be generated once and not every time the position is reset.

JodyLStocks committed:
Updating Delta time to Chrono:
- Fixed the Main loop float delta_time to Chrono Miliseconds delta_time
- Need to fix issue with player collision class. How to Multiply Chrono duration by a scalar value.

JodyLS27 committed:
Update delta time to chono
- Removed delta time as a float and replaced it as an std::chrono time.

JodyLS27 committed:
Fixing if statment mess
- Uncompressed the mess of if statments in player.window_collision()
- Uncompressed the mess of if statments in target.generate_random_position()

JodyLS27 committed:
Fixing Player if statment mess

karnkaul commented on src/main.cpp:
```diff
@@ -1,6 +1,183 @@
+		auto delta_time = current_clock - end_clock;
+		end_clock = current_clock;
+
+		event_handling(window, event, player);
+
+		logic(player, target, window_size, delta_time);
+
+		draw_text(font, text, player.get_score());
+		render(window, player.get_shape(), target.get_shape(), text);
```
Recommend refactoring this logic to be in a class that stores font, text, player, target, etc as member variables, and takes in delta time in its update/tick function.

JodyLS27 replied:
Good Idea, Was going to do this in my next build version with a full class rework.
Happy to do it now though.

karnkaul commented on src/main.cpp:
```diff
@@ -1,6 +1,183 @@
+			{
+			case sf::Keyboard::Up:
+			case sf::Keyboard::W:
+				if (player.can_change_direction("up"))
```
These string comparisons are unnecessary and they are also happening every frame. Use an enum instead.

JodyLS27 replied:
On it, Will Adjust it and move forward, Thanks Karn :)

karnkaul commented on src/player.cpp:
```diff
@@ -0,0 +1,200 @@
+	}
+
+	if (m_move_down && m_move_left)
+	{
+		m_direction = { -0.5f, 1.0f };
+		return;
+	}
+
+	if (m_move_down && m_move_right)
+	{
+		m_direction = { 0.5f, 1.0f };
+		return;
+	}
+
+	// Up, Down, Left, Right Movement
+	if (m_move_up)
+	{
+		m_direction = { 0.0f, -1.0f };
+		return;
+	}
+
+	if (m_move_left)
+	{
+		m_direction = { -1.0f, 0.0f };
+		return;
+	}
+
+	if (m_move_down)
+	{
+		m_direction = { 0.0f, 1.0f };
+		return;
+	}
+
+	if (m_move_right)
+	{
+		m_direction = { 1.0f , 0.0f };
+		return;
+	}
+
+	m_direction = { 0.0f , 0.0f };
```
Why not just update `m_direction` earlier and then normalize it here?

JodyLS27 replied:
I don't quite understand what you mean by earlier?
The `set_direction()` function is the first thing to run after the inputs are mapped and set on that frame.

karnkaul commented on src/player.cpp:
```diff
@@ -0,0 +1,200 @@
+	ret_position = get_position() + (m_direction * dt) * m_speed;
+	//ret_position.x = get_position().y + (m_direction.y * delta_time) * m_speed;
+
+	if (ret_position.x < (m_width * 0.5f))
+	{
+		ret_position.x = (m_width * 0.5f);
+	}
+
+	if (ret_position.x > (window_size.x - (m_width * 0.5f)))
+	{
+		ret_position.x = (window_size.x - (m_width * 0.5f));
+	}
+
+	if (ret_position.y < (m_height * 0.5f))
+	{
+		ret_position.y = (m_height * 0.5f);
+	}
+
+	if (ret_position.y > (window_size.y - (m_height * 0.5f)))
+	{
+		ret_position.y = (window_size.y - (m_height * 0.5f));
+	}
```
Use `std::clamp` instead

JodyLS27 replied:
Yea cool, Better Method will switch to that. Didnt know it was there :)

JodyLS27 replied:
✅ Addressed: Cleaned up and Optimized

karnkaul commented on src/target.hpp:
```diff
@@ -0,0 +1,42 @@
+private:
+	float m_radius{};
+
+	std::default_random_engine m_engine;
```
This isn't seeded with anything, it will generate the same sequence every time. Use `m_engine{std::random_device{}()}` to seed it.

JodyLS27 replied:
Will Adjust this now.

JodyLS27 replied:
✅ Addressed: Moved out of the constructor and Initialized in the header.

JodyLStocks committed:
Updating Member variable
- Adding default values to the member variables in the header
- Removed the m_engine population from the constructor.

JodyLStocks committed:
Adding default header variables values

JodyLStocks committed:
Updated Window collision to be more Optimized

JodyLS27 commented:
Going to close this PR, Too many issues with the code that needs cleanup and Organisation.
I will make a new PR in the Future.
